src/dune_console/dune_console.ml

```diff
@@ -189,33 +249,79 @@ module Backend = struct
       let open T in
       let last = ref (Unix.gettimeofday ()) in
       let frame_rate = 1. /. 60. in
+      let cleanup () =
+        state.finished <- true;
+        Base.finish ();
+        Condition.broadcast finish_cv;
+        Mutex.unlock mutex
+      in
       try
+        Base.start ();
+        (* This is the main event loop for a threaded backend.
+
+           Firstly we lock our mutex, to prevent other threads from mutating our
+           state. Next we ask our implementation to render the given state,
+           afterwards checking if a finish was requested.
+
+           If a finish was requested we exit the loop cleanly.
+
+           We unlock our mutex and go into a time calculation. This calculation
+           gets the current time and compares it with the last recorded time.
+           This lets us compute the elapsed time.
+
+           Next we check that the elapsed time is larger than our specifed
+           [frame_rate]. If this is the case then we can handle any pending user
+           events and continue the loop as soon as possible.
+
+           If we have not yet reached the [frame_rate] then we can handle user
+           events and sleep for the remaining time. *)
         while true do
           Mutex.lock mutex;
-          while not (Queue.is_empty state.messages) do
-            Base.print_user_message (Queue.pop_exn state.messages)
-          done;
-          Base.set_status_line state.status_line;
-          flush stderr;
+          Base.render state;
           let finish_requested = state.finish_requested in
           if finish_requested then raise_notrace Exit;
           Mutex.unlock mutex;
           let now = Unix.gettimeofday () in
           let elapsed = now -. !last in
-          if elapsed >= frame_rate then last := now
-          else
-            let delta = frame_rate -. elapsed in
-            Unix.sleepf delta;
-            last := delta +. now
+          let new_time =
+            if elapsed >= frame_rate then
+              Base.handle_user_events ~now ~time_budget:0.0 mutex
```

Comment:
The `~time_budget:0.0` still bothers me a bit. How can the implementation satisfy it? Surely, it'll spend a little more than zero?

Reply:
> How can an implementation satisfy itHow can an implementation satisfy it

If it doesn't, we'll deviate from the desired frame rate slightly. I've clarified in the comment that `time_budget` is only an estimate and doesn't need to be fully accurate.